### PR TITLE
Detach trait

### PR DESCRIPTION
**Pull request by zzxwill**

### CODE REVIEW

**Comment by zzxwill** [comment]
Merged trait detaching from detach.go to trait_bind in latest commit.

File: pkg/cmd/trait_bind.go [hunk 1]
```diff
@@ -62,12 +65,14 @@ func AddTraitPlugins(parentCmd *cobra.Command, c client.Client, ioStreams cmduti
 		pluginCmd.SetOut(o.Out)
 		for _, v := range tmp.Parameters {
 			pluginCmd.Flags().StringP(v.Name, v.Short, v.Default, v.Usage)
-			if v.Required {
-				pluginCmd.MarkFlagRequired(v.Name)
-			}
+			// These flags are exclusive to `--detach`, so comments the lines.
```
Comment (wonderflow) [inline pkg/cmd/trait_bind.go:68]: according to this case, maybe `<workloadtype>:detach` will be better.

Comment (zzxwill) [reply pkg/cmd/trait_bind.go:68]: > according to this case, maybe `rudr <workloadtype>:detach` will be better.

`rudr <workloadtype>:detach <traittype> -a <applicationName>`?

Comment (wonderflow) [reply pkg/cmd/trait_bind.go:68]: I'm sorry, `rudr <traittype>:detach  applicationName`

Comment (zzxwill) [reply pkg/cmd/trait_bind.go]: Done

File: pkg/cmd/trait_bind.go [hunk 2]
```diff
@@ -118,51 +115,88 @@ func (o *commandOptions) Complete(cmd *cobra.Command, args []string, ctx context
...
+		return err
+	}
+
+	_, _, tKind := cmdutil.GetTraitNameAliasKind(ctx, c, namespace, o.TraitAlias)
+
+	for _, com := range o.AppConfig.Spec.Components {
+		if com.ComponentName == appName {
+			traits := com.Traits
+			traitDefinitionList := cmdutil.ListTraitDefinitionsByApplicationConfiguration(o.AppConfig)
+			for i := 0; i < len(o.AppConfig.Spec.Components[0].Traits); i++ {
```
Comment (wonderflow) [inline pkg/cmd/trait_bind.go:195]: `Components[0]`?

File: pkg/cmd/trait_bind.go [hunk 3]
```diff
@@ -118,51 +115,88 @@ func (o *commandOptions) Complete(cmd *cobra.Command, args []string, ctx context
...
+				if strings.EqualFold(traitDefinitionList[i].Kind, tKind) {
+					o.AppConfig.Spec.Components[0].Traits = append(traits[:i], traits[i+1:]...)
+					i--
+				}
```
Comment (wonderflow) [inline pkg/cmd/trait_bind.go:199]: It feels like we are hacking something here, why don't check one by one and append to a new list?

File: pkg/cmd/trait_bind.go (hunk 2 above)
Comment (zzxwill) [reply pkg/cmd/trait_bind.go]: Done.

File: pkg/cmd/trait_bind.go (hunk 3 above)
Comment (zzxwill) [reply pkg/cmd/trait_bind.go]: Fixed per your suggestions

**Comment by zzxwill** [comment]
Detached two traits from an application, it showed as below.

```
$ rudr ls
NAME       	WORKLOAD             	TRAITS                                                      	STATUS	CREATED-TIME
poc08032042	ContainerizedWorkload	poc08032042-manualscaler-trait,poc08032042-simplerollout-...	False 	2020-08-03 20:43:02 +0800 CST

$ rudr ManualScaler:detach poc08032042
Detaching trait from app 
Succeeded!

$ rudr SimpleRollout:detach poc08032042
Detaching trait from app
Succeeded!

$ rudr ls
NAME       	WORKLOAD             	TRAITS                     	STATUS	CREATED-TIME
poc08032042	ContainerizedWorkload	                           	False 	2020-08-03 20:43:02 +0800 CST
```

**Pull request merged**